crates/e2e/tests/e2e/replace_order.rs

```diff
@@ -0,0 +1,135 @@
+            to_wei(1000)
+        )
+    );
+    tx!(
+        solver.account(),
+        token_b.approve(
+            onchain.contracts().uniswap_v2_router.address(),
+            to_wei(1000)
+        )
+    );
+    tx!(
+        solver.account(),
+        onchain.contracts().uniswap_v2_router.add_liquidity(
+            token_a.address(),
+            token_b.address(),
+            to_wei(1000),
+            to_wei(1000),
+            0_u64.into(),
+            0_u64.into(),
+            solver.address(),
+            U256::max_value(),
+        )
+    );
+
+    // Approve GPv2 for trading
+    tx!(
+        trader_a.account(),
+        token_a.approve(onchain.contracts().allowance, to_wei(15))
+    );
+
+    // Place Orders
+    let services = Services::new(onchain.contracts()).await;
+    services.start_protocol(solver).await;
+
+    let order = OrderCreation {
+        sell_token: token_a.address(),
+        sell_amount: to_wei(10),
+        buy_token: token_b.address(),
+        buy_amount: to_wei(5),
+        valid_to: model::time::now_in_epoch_seconds() + 300,
+        kind: OrderKind::Sell,
+        ..Default::default()
+    }
+    .sign(
+        EcdsaSigningScheme::Eip712,
+        &onchain.contracts().domain_separator,
+        SecretKeyRef::from(&SecretKey::from_slice(trader_a.private_key()).unwrap()),
+    );
+    let order_id = services.create_order(&order).await.unwrap();
+
+    // Replace order
+    let new_order = OrderCreation {
+        sell_token: token_a.address(),
+        sell_amount: to_wei(11),
+        buy_token: token_b.address(),
+        buy_amount: to_wei(3),
+        valid_to: model::time::now_in_epoch_seconds() + 300,
+        kind: OrderKind::Sell,
+        app_data: OrderCreationAppData::Full {
+            full: format!(
+                r#"{{"version":"0.9.0","metadata":{{"replacedOrder":"{}"}}}}"#,
+                order_id
+            ),
+        },
+        ..Default::default()
+    }
+    .sign(
+        EcdsaSigningScheme::Eip712,
+        &onchain.contracts().domain_separator,
+        SecretKeyRef::from(&SecretKey::from_slice(trader_a.private_key()).unwrap()),
+    );
+    services.replace_order(&new_order).await.unwrap();
+
+    // Check the previous order is cancelled
+    let old_order = services.get_order(&order_id).await.unwrap();
+    assert_eq!(old_order.metadata.status, OrderStatus::Cancelled);
+
+    // Drive solution
+    tracing::info!("Waiting for trade.");
+    let balance_before = token_b.balance_of(trader_a.address()).call().await.unwrap();
+    wait_for_condition(TIMEOUT, || async { services.solvable_orders().await == 1 })
+        .await
+        .unwrap();
+
+    wait_for_condition(TIMEOUT, || async { services.solvable_orders().await == 0 })
+        .await
+        .unwrap();
+
+    let balance_after = token_b.balance_of(trader_a.address()).call().await.unwrap();
+    assert!(balance_after.checked_sub(balance_before).unwrap() >= to_wei(3));
```

Comment:
One more comment on this: Initially I thought the `>= 8` should lead to an error but I guess the liquidity pools are configured such that you get loads of surplus. That means using the buy token to verify that only the second trade happened doesn't really work.
What you can do instead is to verify that the new sell token balance is exactly `to_wei(3)` less than at the start. For this to be bullet proof you need to create these orders as fill-or-kill (i.e. `partiallyFillable: false`).

Reply:
The assertion to test that only 1 of the orders got executed still uses the buy token balance instead of the sell token balance.
Using the sell token balance for sell orders is easier to grasp because we know for a fact that exactly this amount should be used. Whereas if we assert something about the buy token balance the exact amounts depend on how much surplus the solution generated.

Reply:
Oh, I misunderstood you. Clear now! thank you!